File: algorithm/c/thinkandcount.c
#include<stdio.h>

typedef struct info{
    int height;
    int width;
}Info;

char board[2000][2005];
int sw_line[2000][2005];
int rm,cm;
Info stack[2005];

void statistic(){
    int i,j;
    int count = 0;
    for(i=0;i<rm;i++){
        count = 0;
        for(j=cm-1;j>-1;j--){
            if(board[i][j] == 'w'){
                count++;
            }else{
                count = 0;
            }
            sw_line[i][j] = count;
        }
    }
}

long long countFrom(int col){
    int i;
    long long total = 0;
    long long add = 0;
    int height = 0;
    int s = 0;
    for(i=0;i<rm;i++){
        height = 1;
        while(sw_line[i][col] < stack[s].width){
            add -= stack[s].width * stack[s].height;
            height += stack[s].height;
            s--;
        }
        add += height * sw_line[i][col];
        total += add;
        s++;
        stack[s].height = height;
        stack[s].width = sw_line[i][col];
    }
    return total;
}

int main(){
    int i;
    long long count = 0;
    stack[0].height = 0;
    stack[0].width = 0;
    while(scanf("%d%d",&rm,&cm) != EOF){
        count = 0;
        for(i=0;i<rm;i++){
            scanf("%s",board[i]);
        }
        statistic();
        for(i=0;i<cm;i++){
            count += countFrom(i);
        }
        printf("%lld\n",count);
    }

}

File: algorithm/python/thinkandcount.py
import time

def statistic():
    for i in range(rm):
        line = [0 for i in range(cm)]
        sw_count = 0
        for j in range(cm-1, -1, -1):
            if board[i][j] is 'w':
                sw_count = sw_count + 1
                line[j] = sw_count
            else:
                sw_count = 0
        sw_line.append(line)

def countFrom(rs,cs):
    total = 0
    re = rm
    ce = sw_line[rs][cs]
    for i in range(rs, re):
        if sw_line[i][cs] < ce:
            ce = sw_line[i][cs]
        total = total + ce
        if ce is 0:
            break
    return total

while True:
    try:
        a = input()
    except:
        break
    rm, cm = [int(i) for i in a.split()]
    board = []
    count = 0
    for i in range(rm):
        row = input()
        board.append(row)
    sw_line = []
    statistic()
    for i in range(rm):
        for j in range(cm):
            part_count = countFrom(i,j)
            count = count + part_count
    print(count)

